Implemented load_client_images in ImageTable.py and wrote associated unit tests

Ayala-Lagunas opened the pull request.

Ayala-Lagunas committed:
Implemented load_client_images function and tests

Added comments to load_client_images test cases

Altered file path in load_client_images test cases

Altered file path in load_client_images test cases

Altered test cases for load_client_images

Altered file path in load_client_images test cases

Altered file path in load_client_images test cases

ptelang commented on cvpy/base/ImageTable.py:
```diff
@@ -406,3 +407,62 @@ def from_table(cas_table: CASTable, image_type: ImageType = None,
+        # Find the image type stored in the table
+        image_type = ImageTable._get_image_type(table)
+
+        # Drop the '_type_' column due to conflict between it and other functions
```
Can you add a comment here saying: "This is a temporary fix and will be removed after the actions are modified to accept varchar _type_ column."

Ayala-Lagunas replied:
Yes I can.

ptelang commented on cvpy/tests/test_imagetable.py:
```diff
@@ -266,6 +266,139 @@ def test_imagetable_from_table_custom_col_names(self):
+                path='path/does/not/exist',
+                connection=self.s,
+                subdirs=False
+            )
```
It seems like you have three test cases under test_imagetable_load_client_images function. We should have one test case per function. 

Separate the test cases into different functions. e.g. 
 
test_imagetable_load_client_images
test_imagetable_load_client_images_invalid_path
...

ptelang commented on cvpy/tests/test_imagetable.py:
```diff
@@ -266,6 +266,139 @@ def test_imagetable_from_table_custom_col_names(self):
+                steps=[{'step':{'stepType':'RESIZE', 'type': 'BASIC', 'width': 200, 'height': 200}}]
+            )
+
+            # Compare results to see if they are affected in the same way
```
Instead of python PIL functions, you can use compareimages action for image comparison.

Ayala-Lagunas replied:
I've updated all the tests to use compareimages instead of PIL functions.

ptelang commented on cvpy/tests/test_imagetable.py:
```diff
@@ -266,6 +266,139 @@ def test_imagetable_from_table_custom_col_names(self):
+                table={'name': 'imgsServer'},
+                casout={'name': 'imgsOutServer', 'replace': True},
+                steps=[{'step':{'stepType':'RESIZE', 'type': 'BASIC', 'width': 200, 'height': 200}}]
+            )
```
Is there a need to resize the images before comparison? If not, remove the above calls to processimages action. You can simply compare the images in the tables: imgsClientDirectory, imgsServer.

Ayala-Lagunas replied:
I intended to test that the images were the same before and after an operation, but that is unnecessary.

ptelang commented on cvpy/tests/test_imagetable.py:
```diff
@@ -266,6 +266,139 @@ def test_imagetable_from_table_custom_col_names(self):
+                assert True in i
+            
+            # Assert that the table created from nonexistent path has no images
+            assert len(self.s.fetchimages('imgsClientEmpty').Images.Image) == 0
```
This assert will need to be moved to the other test function. 
test_imagetable_load_client_images_invalid_path

Ayala-Lagunas replied:
I've created the other test function and moved it over.

ptelang commented on cvpy/tests/test_imagetable.py:
```diff
@@ -266,6 +266,139 @@ def test_imagetable_from_table_custom_col_names(self):
+                connection=self.s,
+                subdirs=False
+            )
+            assert len(self.s.tableinfo(caslib='CASUSER').pop('TableInfo')['Name']) == 1
```
Same comment as earlier. Separate out each test case into its own test function. 

test_imagetable_load_client_images_correct_caslib
test_imagetable_load_client_images_incorrect_caslib
test_imagetable_load_client_images_no_caslib
...

ptelang commented on cvpy/tests/test_imagetable.py:
```diff
@@ -266,6 +266,139 @@ def test_imagetable_from_table_custom_col_names(self):
+        if exists(path):
+            # Images are determined to be similar if the average pixel value of all channels
+            # are less than three percent of their maximum value.
+            validThreshold = 255 * 0.03
```
If you use compareimages action (see my comment below), this can be removed.

ptelang commented on cvpy/tests/test_imagetable.py:
```diff
@@ -266,6 +266,139 @@ def test_imagetable_from_table_custom_col_names(self):
+        from os.path import exists
+
+        # Path to the directory the function will load images from
+        path = "images/"
```
This should be the complete path to the images directory. 

path = f"{TestImageTable.DATAPATH}/data/images"

ptelang reviewed (commented):
Hi Andres, Great job with this change. Please see my comments below.

Ayala-Lagunas commented on cvpy/tests/test_imagetable.py:
```diff
@@ -266,6 +266,139 @@ def test_imagetable_from_table_custom_col_names(self):
+        path = "dicom/"
+
+        # If path exists, we are on the Jenkins server
+        if exists(path):
```
This was intended to see if the path exists on the Jenkins server to differentiate between local testing and Jenkins testing, but I've removed it.

Ayala-Lagunas commented on cvpy/base/ImageTable.py:
```diff
@@ -406,3 +407,62 @@ def from_table(cas_table: CASTable, image_type: ImageType = None,
+        connection.altertable(table=output_table_parms['name'], columns=[{'name': '_image_', 'binaryType': 'image'}])
+
+        # Print a message stating how many images were added from the path and where they are being stored
+        amount = len(connection.fetchimages(output_table_parms['name']).Images.Image)
```
Since I've switched to using compareimages to test if the images are the same, I've removed this code and some of what came after as well.

Ayala-Lagunas committed:
Altered test cases for load_client_images

courtneyambrozic committed:
Update test_imagetable.py

In this commit, the test file parameters have been updated to included a local path that contains the images needed for testing.

Ayala-Lagunas committed:
Increased load_client_images data type support.

Ayala-Lagunas committed:
Merge branch 'sassoftware:main' into main

Ayala-Lagunas committed:
Merge branch 'main' of https://github.com/Ayala-Lagunas/python-cvpy

Ayala-Lagunas committed:
ImageTable now allows char and varchar types.

Ayala-Lagunas committed:
Removed load_client_images biomed image test case.

courtneyambrozic reviewed (approved):
Looks great, Andres! Thanks for your work on this.
